_sips/sips/2017-09-20-opaque-types.md

```diff
@@ -736,6 +753,178 @@ This example is somewhat similar to `Logarithm` above. Other
+to add extra type safety to a "stringly-typed" API, by replacing
+instances of the `String` type with a more specific type.
+
+For example:
+
+```scala
+package object pkg {
+
+  import Character.{isAlphabetic, isDigit}
+
+  class Alphabetic private[pkg] (val value: String) extends AnyVal
+
+  object Alphabetic {
+    def fromString(s: String): Option[Alphabetic] =
+      if (s.forall(isAlphabetic(_))) Some(new Alphabetic(s))
+      else None
+  }
+
+  opaque type Digits = String
+
+  object Digits {
+    def fromString(s: String): Option[Digits] =
+      if (s.forall(isDigit(_))) Some(s)
+      else None
+
+    def asString(d: Digits): String = d
+  }
+}
+```
+
+The code here is relatively comparable. However, when changing
+`String` to `Alphabetic` in code, the following types would be changed
+(i.e. boxed):
+
+ * `Array[Alphabetic]`
+ * `Option[Alphabetic]` (e.g. the result of `Alphabetic.fromString`)
+ * `Vector[Alphabetic]`
+ * `Alphabetic => Boolean`
+ * `Map[Alphabetic, Int]`
+ * `Ordering[Alphabetic]`
+ * `(Alphabetic, String)`
+ * `Monoid[Alphabetic]`
+
+In many cases users won't mind the fact that this code will box, but
+there will certainly be an impact on the bytecode produced (and
+possibly the runtime performance).
+
+By contrast, replacing `String` with `Digits` is guaranteed to have no
+impact (all occurances of `Digits` are guaranteed to be erased to
+`String`). Aside from the ergonomics of calling the `fromString` and
+`asString` methods, there's no runtime impact of any kind.
+
+### Default visibility
+
+By default, a value class' constructor and accessor are public. It
+*is* possible to restrict access, using a `private` constructor and
+`private` accessor, but this makes the comparison between opaque types
+and value classes less attractive:
+
+```scala
+package object pkg {
+  opaque type XCoord = Int
+
+  case class private[pkg] YCoord (private[pkg] val Int) extends AnyVal
+
+  // in both cases, we'd need public factory constructors
+  // to allow users to produce values of these types.
+}
+```
+
+### LUB differences
+
+Value classes extend `AnyVal` by virtue of the syntax used to define
+them. One reason this is necessary is that value classes cannot be
+`null` (otherwise this could create ambiguities between their boxed
+and unboxed representations when wrapping `AnyRef` values).
+
+By contrast, when seen from the "outside" opaque types extend
+`Any`. Their bounds are the same as those of a type parameter without
+explicit bounds (e.g. `A <: Any >: Nothing`).
+
+### Size of artifacts produced
+
+Since value classes do have a runtime reprsentation, they do increase
+the size of runtime artifacts produced (whether a JAR file, or a
+javascript file). Their methods are also compiled to multiple
+representations (i.e. they support both the boxed and unboxed forms
+via extensions methods). Again, this comes at a cost.
+
+By contrast, opaque types have no inherent runtime footprint. The
```

Comment:
An example might be useful here. That said, maybe this is actually overstated. If we have methods they have to go somewhere so that will be a class, and so really we are arguing for smaller code. I think it could be true, but it is not 100% obvious without knowing the full encoding. Lastly, I do think reduction of compiled code side is important for scala. It is not uncommon to see it as a complaint and some say one of the reasons scala has not been popular for android developers.